File: backend/app/models/scores.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class DimensionScores(BaseModel):
    """Individual dimension scores (0-100)."""

    p_time: int = Field(..., ge=0, le=100, description="Schedule adherence score")
    p_cost: int = Field(..., ge=0, le=100, description="Budget adherence score")
    p_quality: int = Field(..., ge=0, le=100, description="Product quality score")
    p_value: int = Field(..., ge=0, le=100, description="Strategic value score")
    p_satisfaction: int = Field(..., ge=0, le=100, description="Client satisfaction score")
    p_flow: int = Field(..., ge=0, le=100, description="Flow & predictability score")
    p_engineering: int = Field(..., ge=0, le=100, description="Engineering maturity score")
    p_risk: int = Field(..., ge=0, le=100, description="Risk posture score")


class DoraScore(BaseModel):
    """DORA metrics score (separate from main score)."""

    score: int | None = Field(None, ge=0, le=100, description="DORA score 0-100")
    classification: str | None = Field(None, description="Elite/High/Medium/Low")
    metrics: dict[str, float | None] = Field(
        default_factory=dict,
        description="Individual DORA metric scores (0-1)",
    )
    available_metrics: int = Field(0, description="Number of DORA metrics with data")


class FinalScore(BaseModel):
    """Final weighted aggregate score."""

    score: int = Field(..., ge=0, le=100)
    dimensions: DimensionScores
    weights_applied: dict[str, float]
    dora: DoraScore | None = Field(None, description="Separate DORA score")


class ScoreResult(BaseModel):
    """Complete scoring result."""

    id: UUID
    project_id: UUID
    indicators_id: UUID
    dimensions: DimensionScores
    final_score: int = Field(..., ge=0, le=100)
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/modules/scorecard/types/scores.ts
export interface DimensionScores {
  p_time: number | null;
  p_cost: number | null;
  p_quality: number | null;
  p_value: number | null;
  p_satisfaction: number | null;
  p_flow: number | null;
  p_engineering: number | null;
  p_risk: number | null;
}

export type Dimension = 'Time' | 'Cost' | 'Quality' | 'Value' | 'Satisfaction' | 'Flow' | 'Engineering' | 'Risk';

export const ALL_DIMENSIONS: Dimension[] = ['Time', 'Cost', 'Quality', 'Value', 'Satisfaction', 'Flow', 'Engineering', 'Risk'];

export interface HistoricalDataPoint {
  period: string;
  value: number | null;
}

export type DoraLevel = 'Elite' | 'High' | 'Medium' | 'Low';

export interface DoraMetricDetail {
  value: number;
  level: DoraLevel;
  score: number;
  no_incidents?: boolean;
}

export interface DoraScore {
  score: number | null;
  classification: DoraLevel | null;
  metrics: {
    deployment_frequency?: DoraMetricDetail;
    lead_time?: DoraMetricDetail;
    change_failure_rate?: DoraMetricDetail;
    mttr?: DoraMetricDetail;
  };
  available_metrics: number;
}

export interface FinalScore {
  score: number | null;
  dimensions: DimensionScores;
  weights_applied: Record<string, number>;
  dora: DoraScore | null;
}

export interface Indicators {
  spi: number | null;
  on_time_milestones: number | null;
  cpi: number | null;
  budget_variance: number | null;
  defect_density: number | null;
  escaped_rate: number | null;
  mttr_hours: number | null;
  governance_compliance: number | null;
  lead_time_days: number | null;
  commitment_reliability: number | null;
  pr_review_ratio: number | null;
  prs_without_review: number | null;
  high_vulns: number | null;
  test_maturity: number | null;
  arch_checklist: number | null;
  story_review_ratio: number | null;
  okr_impact: number | null;
  pm_satisfaction: number | null;
  client_satisfaction: number | null;
  pr_size_median: number | null;
  review_turnaround_hours: number | null;
  deployment_frequency: number | null;
  change_failure_rate: number | null;
  post_contract_tasks: number | null;
}

export interface ScoreResponse {
  indicators: Indicators;
  scores: FinalScore;
}

export interface ScoringConfig {
  targets: {
    defect_density: number;
    escaped_rate: number;
    mttr_hours: number;
    spi: number;
    cpi: number;
    lead_time_days: number;
    high_vuln_count: number;
    gov_exceptions: number;
    pr_no_review_ratio: number;
    story_review_ratio: number;
    client_satisfaction: number;
    architecture: number;
    commitment_reliability: number;
    milestones_on_time: number;
    test_maturity: number;
    pm_satisfaction: number;
    pr_size_lines: number;
    review_turnaround_hours: number;
    deployment_frequency: number;
    change_failure_rate: number;
    post_contract_tasks: number;
    budget_variance: number;
    governance_compliance: number;
    okr_impact: number;
  };
  ideals: {
    spi: number;
    cpi: number;
  };
  global_weights: {
    time: number;
    cost: number;
    quality: number;
    value: number;
    satisfaction: number;
    flow: number;
    engineering: number;
    risk: number;
  };
  constants: {
    sev1_cap: number;
    grace_days: number;
  };
  weight_validation: Record<string, boolean>;
}
